src/internal/auth/handler.go

```diff
@@ -5,47 +5,63 @@ import (
 	"github.com/pachyderm/pachyderm/v2/src/client"
 )
 
-type authHandler func(*client.APIClient, string) error
+// an authHandler can optionally return a key-value that will be cached in the request's context
+type authHandler func(*client.APIClient, string) (*keyValue, error)
```

Comment:
Returning an arbitrary key-value pair might be overkill - I think we could just return the username as a string for now to make this a little simpler.